### PR TITLE
[Kubernetes - Metricbeat] Providing config option to disable Kubeadm config api requests

### PR DESCRIPTION
**Pull request by gizas**
- Enhancement

~**NOTE:** Dont merge until https://github.com/elastic/elastic-agent-autodiscover/pull/98 is also merged~

The elastic-agent-autodoscover library updated to [v0.8.2](https://github.com/elastic/beats/pull/40086/files#diff-3295df7234525439d778f1b282d146a4f1ff6b415248aaac074e8042d9f42d63R559)

## Proposed commit message

**WHAT**: Provide a configuration option to disable `/api/v1/namespaces/kube-system/configmaps/kubeadm-config` . In more details user will disable the api requests to the modules he configures by using configuration `use_kubeadm: false`

**WHY**: https://github.com/elastic/enhancements/issues/21172

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Install a kubernetes cluster with audit-logs enabled
- Create a local elastic stack with `elastic-package stack up -d -v --version=8.15.0-SNAPSHOT`
- Update the go.mod file of this repo and point to the elastic-agent-autodiscovery folder, where related changes should also exist. Your go.mod file should have the following entry: 
```bash
replace github.com/elastic/elastic-agent-autodiscover => /Users/andreasgkizas/elastic/elastic-agent-autodiscover
```
- Run `go mod tidy`
- Install Filebeat and collect audit logs
- Follow instructions of this [page](https://github.com/elastic/beats/blob/main/metricbeat/module/kubernetes/_meta/test/docs/README.md) to build the metricbeat binary
```bash
 GOOS=linux GOARCH=arm64 go build
 kubectl apply -f module/kubernetes/_meta/test/docs/01_playground/metricbeat.yaml
 kubectl cp metricbeat `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'`:/usr/share/metricbeat/ -n kube-system\n
 kubectl exec `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'` -n kube-system -- bash -c "metricbeat -e -c /etc/metricbeat.yml"
```
- Use `use_kubeadm: false` flag in your enabled modules
- Open your local kibana cluster, navigate to Discovery and search inside kibana audit-logs for relevant api requests
Screenshot 2024-07-03 at 12 06 49 PM

## Related issues

-  Relates #https://github.com/elastic/enhancements/issues/21172
- Relates #https://github.com/elastic/elastic-agent-autodiscover/pull/98

### With 8.15.0 Image no changes:

Screenshot 2024-07-03 at 12 06 49 PM

We enable all metricsets (included events) and we have 85 API requests

### With This repo code but the flag is `disable_kubeadm: false`

Screenshot 2024-07-03 at 12 34 26 PM

We enable all metricsets (included events) and we still have 85 API requests

### With This repo code but the flag is disable_kubeadm: true`

Screenshot 2024-07-03 at 1 08 16 PM

We have no API requests after 10:04 that metrics collection established

```json
{"log.level":"info","@timestamp":"2024-07-03T10:04:20.520Z","log.logger":"publisher_pipeline_output","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*netClientWorker).run","file.name":"pipeline/client_worker.go","file.line":145},"message":"Connection to backoff(elasticsearch(https://elasticsearch:9200)) established","service.name":"metricbeat","ecs.version":"1.6.0"}
```

### CODE REVIEW

**Comment by gizas** [comment]
I have updated description and autodiscovery library to make use of use_kubeadm variable

See comment https://github.com/elastic/elastic-agent-autodiscover/pull/98#discussion_r1684158494

File: metricbeat/module/kubernetes/util/kubernetes.go [hunk 1]
```diff
@@ -42,9 +42,9 @@ import (
 type kubernetesConfig struct {
 	KubeConfig        string                       `config:"kube_config"`
 	KubeClientOptions kubernetes.KubeClientOptions `config:"kube_client_options"`
-
-	Node       string        `config:"node"`
-	SyncPeriod time.Duration `config:"sync_period"`
+	KubeAdm           bool                         `config:"use_kubeadm"`
```
Comment (MichaelKatsoulis) [inline metricbeat/module/kubernetes/util/kubernetes.go:45]: I don't think it is needed here as long as it is in https://github.com/elastic/elastic-agent-autodiscover/blob/a1dcbcec6e6b14f4a826d3a07c2eb93d9412907f/kubernetes/metadata/config.go#L27

File: metricbeat/module/kubernetes/util/kubernetes.go [hunk 2]
```diff
@@ -527,6 +528,9 @@ func createMetadataGenSpecific(client k8sclient.Interface, commonConfig *conf.C,
 		}
 		// For example for pod named redis in namespace default, the generator uses the pod watcher for pod metadata,
 		// collects all node metadata using the node watcher's store and all namespace metadata using the namespacewatcher's store.
+		// We initialise the use_kubeadm variable based on modules KubeAdm base configuration
```
Comment (MichaelKatsoulis) [inline metricbeat/module/kubernetes/util/kubernetes.go:531]: We could set this before the call of `createMetadataGenSpecific` and `createMetadataGen`

Comment (gizas) [reply metricbeat/module/kubernetes/util/kubernetes.go]: Thanks done in here https://github.com/elastic/beats/pull/40086/files#diff-1d7b17f1ef239e6177db5ee660f2234bfd69c82e93494d4e46ee97f8a8617cfdR595

**Comment by gizas** [comment]
@MichaelKatsoulis I added again the use_kubeadm option [here](https://github.com/elastic/beats/pull/40086/files#diff-c00c194167e320d0bb1f2291fe85b8eb12853ffa5074e633e06b2d607b625845R28)

Reason is that kubelet metricsets, proxy and scheduler make use of AddClusterECSMeta (like [here](https://github.com/elastic/beats/blob/main/metricbeat/module/kubernetes/apiserver/metricset.go#L48)) which is using finally the [kubernetes config](https://github.com/elastic/beats/blob/main/metricbeat/module/kubernetes/util/kubernetes.go#L42) struct. 

I have repeated the tests and now works ok!

**Review by MichaelKatsoulis — APPROVED** [review]
LGTM

**Comment by gizas** [comment]
@elastic/beats-tech-leads can I have a review here? The pr looks good and tested

**Review by cmacknz — APPROVED** [review]
Approving go.mod changes.

**Review by tetianakravchenko — APPROVED** [review]
LGTM!

**Pull request merged**